### PR TITLE
[feature request] : added python script to automate the setup of talawa-api

### LINKED ISSUE #885: [Feature Request] : making a python script to automate the setup of talawa-api
### Is your feature request related to a problem?
For anyone who wishes to contribute to talawa, setting up a local instance of the talawa-api is one of the most crucial steps. I personally feel this is a step where many people either get stuck for a long time or loose the interest to debug it further.

### Proposed solution
I propose to make a shell script (.sh file) that will automate most of the setup for talawa-api. Although every single part may not be suitable for automation, but most of the part like setting up access tokens, refresh tokens, generating a local MongoDB url can be done error-free within seconds, without any hassle.

### Alternatives considered
As of now, I think that making a shell script should solve the problem. But if for some reason it seems to not be feasible in the future, I think an alternative would be a python script to automate the process.

### Approach
My approach would be making a shell script. To run this, all the user has to do is run the following command : - ```./installer.sh```
This will generate a new .env file, fill it with autogenerated access and token secrets, and a MongoDB url for the ones who prefer to use a local instance. For steps that can’t be automated, such as generating a Google recaptcha secret key etc, instructions can be shown on how to proceed.
For steps where configurations may be based on user preference, such as using a cloud/local instance of MongoDB or the preferred PORT for the api to be run on, multiple options can be provided in the cli itself.

### PR DESCRIPTION
**Pull request by rv602**
### What kind of change does this PR introduce?
This PR is a feature addition. I have added a python script in the root directory of the project which will help in a much quicker and efficient way of setting up talawa-api for developers.

### Issue Number: 

Fixes #885 

### Video : -

https://user-images.githubusercontent.com/92572013/215287412-45c40b86-17fd-43e0-b64d-7d534714ed84.mp4

### Did you update the documentation?
Yes.
Just added a line in README.md saying that developers can run the command python3 installer.py to quickly setup talawa-api

### Summary : -
The motivation to raise this feature request and work on it was convenience and saving time of developers who wish to contribute to talawa. Many people make minor errors in a hassle to setup a project and that leads them to unnecessary errors and eventually loosing motivation to debug the errors further. 
This script aims to eliminate the possibility of almost any minor errors possible from the developer's side.

### Does this PR introduce a breaking change?
No

### Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?
Yes

### CODE REVIEW

**Review by noman2002 — DISMISSED** [review]
Looks good to me. Good work.

**Comment by rv602** [comment]
@noman2002 @palisadoes workflows are successful. Code coverage has also been generated and is 100%
So I guess this can be merged now 😅.

**Comment by palisadoes** [comment]
This is a good start, but I'd ask you to go the extra mile.

We have this issue with Admin setup. Organizations and super users are not automatically setup in MongoDB.

https://github.com/PalisadoesFoundation/talawa-admin/issues/414

Please also add this functionally to the script.

It should not crash if MongoDB isn't running, but give a useful message.

Please run python black against your code. Also run pylint and pydocstyle against it too until there are no errors.

This will be a great improvement. Thanks in advance for the work. 

This may take a few days because of the importance of what we are trying to achieve.

Please also think of ways of writing the script so that we can run tests against some of the key functions. This is because of we are using it to update the database we'll need to make sure it always works as expected

**Comment by rv602** [comment]
@palisadoes, even I had that thought of automating the setup of a Superadmin, but didn’t implement it here as I felt I could implement it in the talawa-admin repository.
But as you mentioned, I feel it would be more feasible to add this part to this python script itself.
I will make a commit soon with the necessary changes.
Thank you.

**Comment by rv602** [comment]
You can assign me this issue in the talawa-admin repository as well
https://github.com/PalisadoesFoundation/talawa-admin/issues/414

I will make the necessary changes there in the instructions after I'm done with the python script

**Comment by noman2002** [comment]
> You can assign me this issue in the talawa-admin repository as well [PalisadoesFoundation/talawa-admin#414](https://github.com/PalisadoesFoundation/talawa-admin/issues/414)
> 
> I will make the necessary changes there in the instructions after I'm done with the python script

Please comment of that particular issue.  It will be assigned to you.

**Comment by rv602** [comment]
okay I'll do it now. Thank you

**Comment by rv602** [comment]
@palisadoes @noman2002 I have made the necessary changes.

### Video of the newer version of the script : -

https://user-images.githubusercontent.com/92572013/215339626-720adc13-473e-4c71-93a8-7398d64ffe6c.mp4

### Running black against python script : -

image

### Running pylint against the script : -

image

### Pending work : -
There are a few minor changes I've to do in the script, and running it against pydocstyle is remaining.

I'll make a new commit shortly with all the changes

**Comment by rv602** [comment]
@palisadoes @xsh2047 my latest commit contains all changes that had been asked.

### Snapshot of black, pylint and pydocstyle run against the script : -

image

### Video of the latest script : -

https://user-images.githubusercontent.com/92572013/215408316-4b9fb322-e62b-4569-8cd1-ee4d34553eb6.mp4

**Review by palisadoes — CHANGES_REQUESTED** [review]
1. Please remove the colored output from the script
1. Only put imports in the requirements file for packages that you use in the script.
1. Use configparser for reading the `.env` file. It will be easier.
1. What messages are being ignored in `.pylintrc`?

File: installer.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+                with open(".env", "r", encoding="utf-8") as file:
+                    lines = file.readlines()
+                replace_line(
+                    ".env",
+                    [i for i, line in enumerate(lines) if "MONGO_DB_URL=" in line][0],
+                    f"MONGO_DB_URL={cloud_instance}\n",
+                )
+                db_url = cloud_instance
+            else:
+                print(
```
Comment (palisadoes) [inline installer.py:240]: Make the mongodb interaction a separate function

File: installer.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    new_user = {
+        "firstName": "test",
+        "lastName": "admin acc",
+        "email": "test@email.com",
+        "password": "Pass@1234",
+        "userType": "SUPERADMIN",
+        "tokenVersion": 0,
+        "appLanguageCode": "en",
+        "createdOrganizations": [],
+        "createdEvents": [],
```
Comment (palisadoes) [inline installer.py:269]: 1. Does this test whether the user already exists?
2. Does this test whether the user already has the privileges?

**Review by palisadoes — CHANGES_REQUESTED** [review]
There is room for improvement. Please take a look at the comments. Let's reduce the risks of a poor UX.

File: installer.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
+"""This script is used to install Talawa API on your local machine."""
```
[the remark this answers is not in the document]
Comment (rv602) [reply installer.py:1]: Okay I'll specify this here and in the instructions as well.

File: installer.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+import datetime
+import os
+import base64
+import secrets
+import subprocess
+import sys
+import pymongo
+from pymongo import MongoClient
+
+TGREEN = "\033[32m"
```
[the remark this answers is not in the document]
Comment (rv602) [reply installer.py:11]: Okay I'll remove the colours from everywhere.

File: installer.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        curr_lines = file.readlines()
+    curr_lines[line_num] = text
+
+    with open(file_name, "w", encoding="utf-8") as file:
+        file.writelines(curr_lines)
+
+
+def env_creation():
+    """Create a '.env' file by copying '.env.sample' file."""
+    os.system("cp .env.sample .env")
```
[the remark this answers is not in the document]
Comment (rv602) [reply installer.py:42]: I guess that ```os.system``` should work as desired on Windows as well. I'll test it and get back on this.

File: installer.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+
+def access_and_refresh_tokens():
```
[the remark this answers is not in the document]
Comment (rv602) [reply installer.py:45]: I haven’t used ```configparser``` before 😅.
I'll look into it and try implementing it in the script

File: installer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rv602) [reply installer.py:42]: Okay I'll try implementing this feature

File: installer.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+        print(TRED + "\nConnection to MongoDB failed. Please try again.", TWHITE)
+        print("\nError: ", error, TWHITE)
+        abort()
+        return False
+
+
+def recaptcha():
+    """Request user to set up reCAPTCHA and updates the secret key in '.env' file."""
+    print(
+        TGREEN
```
[the remark this answers is not in the document]
Comment (rv602) [reply installer.py:94]: Alright I'll check it

File: installer.py [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+        + "and under Signing in to Google section select App Passwords."
+    )
+    print(
+        "Click on Select app section and choose Other(Custom name), enter talawa as the custom name and press Generate button."
+    )
+
+    email = input("\nEnter your email: ")
+    password = input("Enter the generated password: ")
+
+    with open(".env", "r", encoding="utf-8") as file:
```
[the remark this answers is not in the document]
Comment (rv602) [reply installer.py:142]: Okay I'll look into it

File: installer.py [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+        file.writelines(content)
+
+
+def npm_install():
+    """Installs the dependencies using npm."""
+    print("\nInstalling dependencies... 📦")
+
+    with open("npm_install.log", "w", encoding="utf-8") as log_file:
+        npm_result = subprocess.run(
+            "npm install",
```
[the remark this answers is not in the document]
Comment (rv602) [reply installer.py:168]: In the INSTALLATION.md, the command given is ```npm install```.
I can maybe try giving the user a few choices and see which package manager is the preferred one and run the corresponding command to install dependencies.
If the user uses a package manager that isn’t listed in the options, we can maybe ask him to type the equivalent command to install packages and run the command as he gives it.

File: installer.py (hunk 1 above)
Comment (rv602) [reply installer.py:240]: Yes I'll do it

File: installer.py [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+    """Notifies user to set up Two-Factor Authentication on their Google Account."""
+    print(TRED + "\nIMPORTANT ❗️", TWHITE)
+    print(
+        "\nEnsure that you have Two-Factor Authentication set up on your Google Account. Visit this URL:"
+    )
+    print(
+        TBLUE
+        + "\nhttps://myaccount.google.com"
+        + TWHITE
+        + "\n\nselect"
```
[the remark this answers is not in the document]
Comment (rv602) [reply installer.py:129]: Okay will do

File: installer.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (palisadoes) [reply installer.py:168]: Please finish on configuring the DB and the `.env` file

File: installer.py (hunk 2 above)
Comment (rv602) [reply installer.py:269]: Well as of now it doesn’t check anything of that sort exists in the DB.
I'll add a check for that, and if a SUPERADMIN account already exists I'll just display its credentials.

File: installer.py [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+        )
+        abort()
+
+    os.remove("npm_install.log")
+    print("Dependencies installed! 📦")
+
+
+def abort():
+    """Aborts the installation process and deletes the '.env' file."""
+    os.remove(".env")
```
[the remark this answers is not in the document]
Comment (palisadoes) [reply installer.py:190]: > understand

Don't create an `.env` file if one did not exist before.

**Comment by palisadoes** [comment]
@rv602 Are you still working on this?

**Comment by rv602** [comment]
@palisadoes, yes I am still working on it, I'll make the necessary changes and make a new commit in a few days.

**Comment by EshaanAgg** [comment]
I had an idea to extend the functionality of the same python script.
The brief overview of my idea is provided [here](https://github.com/PalisadoesFoundation/talawa-admin/issues/433).

We can add the option to create these objects for `talawa-admin` as the last step of the script.

**Comment by rv602** [comment]
@palisadoes I am still working on the script.
Actually I have my University end-semester exams in the upcoming days, due to which there may be a slight delay from me in my work for a few days. Kindly excuse me for the same.
Thank you.

**Comment by palisadoes** [comment]
@rv602 I recognize the interruption caused by exams, but we also need to plan. This is an important PR 
- When do you think you'll be able to restart on this?
- When this is completed there is also that needs to be worked on https://github.com/PalisadoesFoundation/talawa-api/issues/1012. I've assigned it to myself for now. Please do not add these features to this script.

**Comment by rv602** [comment]
@palisadoes my main exams will be till 17th February.
I will try to make minor contributions as much as I can till then.
After 17th February I will resume my work properly again.

File: installer.py (hunk 2 above)
Comment (rv602) [reply installer.py:269]: @palisadoes if a ```SUPERADMIN``` already exists, what should be the flow of the script?
Should I ask the user if he would like to create a new ```SUPERADMIN``` or not create a new ```SUPERADMIN``` bat all?

Comment (palisadoes) [reply installer.py:269]: It should just mention that one is already found.

Comment (rv602) [reply installer.py:269]: okay

File: installer.py [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+    two_factor_auth()
+
+    client = MongoClient(db_url)
+    db = client["talawa-api"]
+    users_collection = db["users"]
+
+    new_user = {
+        "firstName": "test",
+        "lastName": "admin acc",
+        "email": "test@email.com",
```
Comment (palisadoes) [inline installer.py:263]: They should be prompted for
FirstName
LastName
Email
Password

**Comment by EshaanAgg** [comment]
Wouldn't it be better we abstract the database operations like the creation of the `superadmin` into a separate file, which can be just called by `installer.py`?
This would also readability of our code, and later allow us to expand the functionality of the script to add `starter database objects` in a more modular manner.

File: installer.py (hunk 12 above)
Comment (rv602) [reply installer.py:263]: okay I'll do it

File: installer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rv602) [reply installer.py:42]: @palisadoes sorry for the confusion from my side. I tested `os.system` on Windows and it didn't work.
Hence I am using `shutil` instead now as follows : -

`shutil.copy(".env.sample", ".env")`

It works on Windows as desired.
I just wanted to confirm whether using `shutil` okay

[the remark this answers is not in the document]
Comment (palisadoes) [reply installer.py:42]: 1. The script must be OS independent
1. Also read up on using `os.sep` when necessary

**Comment by rv602** [comment]
@palisadoes I have incorporated most of the changes. There are some errors occuring when configparser is being used to write to the .env file, apart from that few final touches are remaining. I will make the updated PR shortly.

**Comment by rv602** [comment]
@palisadoes I have completed making all the changes as asked. The only remaining thing is to get the `pylint` rating from `9.57/10` to `10/10`.
I'll complete this in a few hours and make the updated commit

**Comment by rv602** [comment]
@palisadoes I'm getting some `unauthorised file changes` error by the workflows.

**Comment by rv602** [comment]
@palisadoes actually the changes made in the `.env.sample` are necessary for the configparser to read and write the environment variables as desired. I've just added the section headers in the .env.sample.

**Comment by palisadoes** [comment]
1. If that is necessary it will cause the application to break. Is there another way to define the sections without  `[brackets]`? Preferably with strings that are interpreted as comments by typescript?
2. If not, configparser may not be the best approach. There may be a package specifically written for reading `.env `files. Please research this.

**Comment by rv602** [comment]
@palisadoes I have an idea that could work.
The main issue is defining the `sections` using brackets in the `.env.sample` file right, so can I try modifying the script in such a way that when the execution of the script begins, it will start adding the `sections` in the `.env.sample` file, and when the script is completed, it will remove the sections from the `.env.sample` file.
In this way it will be modified only while the script is being executed

**Comment by palisadoes** [comment]
I did a google search for `python env package` and this came up as the second result. Please do suitable research

1. https://pypi.org/project/python-dotenv/

**Comment by palisadoes** [comment]
I'm going to close this Please close this PR until you have a solution that is ready to be reviewed. It's been open for over a month.

**Comment by rv602** [comment]
@palisadoes apologies for the late reply, I looked into the package you mentioned.
The issue with that package is that once you set values to the environment variables using it, those environment variables are only valid for the duration of the Python script's execution and are not persisted after the script exits.

Hence I was suggesting that I could write the `sections []` while the script is running, and once the script completes the installation process, it will revert the .env.sample back to it's original state.
This way, even the `unauthorised file changes` won’t be there anymore.

**Pull request closed without merging**